File: wear/src/main/java/ca/joshstagg/slate/SlateWatchFaceService.java
package ca.joshstagg.slate;

import android.app.PendingIntent;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.icu.util.Calendar;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.support.v4.content.ContextCompat;
import android.support.wearable.complications.ComplicationData;
import android.support.wearable.complications.ComplicationHelperActivity;
import android.support.wearable.complications.ComplicationText;
import android.support.wearable.complications.SystemProviders;
import android.support.wearable.watchface.CanvasWatchFaceService;
import android.support.wearable.watchface.WatchFaceStyle;
import android.text.TextUtils;
import android.util.Log;
import android.util.SparseArray;
import android.view.Gravity;
import android.view.SurfaceHolder;

/**
 * Slate ca.joshstagg.slate
 * Copyright 2017 Josh Stagg
 */
public class SlateWatchFaceService extends CanvasWatchFaceService {
    private static final String TAG = "SlateWatchFaceService";

    @Override
    public Engine onCreateEngine() {
        Context context = this.getApplicationContext();
        ConfigManager configService = Slate.getInstance().getConfigService();
        return new Engine(context, configService);
    }

    private class Engine extends CanvasWatchFaceService.Engine {

        private final Context mContext;
        private final ConfigManager mConfigService;
        private final Handler mUpdateTimeHandler;
        private final float[][] mTicks = new float[12][];

        private SlatePaints mPaints;
        private SlateTime mSlateTime;
        private Bitmap mBackgroundBitmap;
        private Bitmap mBackgroundScaledBitmap;
        /**
         * Whether the display supports fewer bits for each color in ambient mode. When true, we
         * disable anti-aliasing in ambient mode.
         */
        private boolean mLowBitAmbient;
        private boolean mBurnInProtection;

        private int mWidth;
        private int mHeight;
        private int mComplicationsY;

        private SparseArray<ComplicationData> mActiveComplicationDataSparseArray;

        Engine(Context context, ConfigManager slateConfigService) {
            mContext = context;
            mConfigService = slateConfigService;
            mUpdateTimeHandler = new EngineHandler(this);
        }

        /**
         * Initialize the watch face
         *
         * @param holder SurfaceHolder that the watch face will use
         */
        @Override
        public void onCreate(SurfaceHolder holder) {
            super.onCreate(holder);
            setWatchFaceStyle(new WatchFaceStyle.Builder(SlateWatchFaceService.this)
                    .setStatusBarGravity(Gravity.CENTER_HORIZONTAL | Gravity.TOP)
                    .build());

            mSlateTime = new SlateTime(mContext);
            mPaints = new SlatePaints();

            Drawable backgroundDrawable = ContextCompat.getDrawable(mContext, R.drawable.bg);
            mBackgroundBitmap = ((BitmapDrawable) backgroundDrawable).getBitmap();

            initializeComplication();
        }

        private void initializeComplication() {
            mActiveComplicationDataSparseArray = new SparseArray<>(Constants.COMPLICATION_IDS.length);
            setDefaultSystemComplicationProvider(Constants.LEFT_DIAL_COMPLICATION, SystemProviders.WATCH_BATTERY, ComplicationData.TYPE_SHORT_TEXT);
            setDefaultSystemComplicationProvider(Constants.RIGHT_DIAL_COMPLICATION, SystemProviders.DATE, ComplicationData.TYPE_SHORT_TEXT);
            setActiveComplications(Constants.COMPLICATION_IDS);
        }

        @Override
        public void onPropertiesChanged(Bundle properties) {
            super.onPropertiesChanged(properties);
            mLowBitAmbient = properties.getBoolean(PROPERTY_LOW_BIT_AMBIENT, false);
            mBurnInProtection = properties.getBoolean(PROPERTY_BURN_IN_PROTECTION, false);
        }

        /*
         * Called when there is updated data for a complication id.
         */
        @Override
        public void onComplicationDataUpdate(int complicationId, ComplicationData complicationData) {
            // Adds/updates active complication data in the array.
            mActiveComplicationDataSparseArray.put(complicationId, complicationData);
            invalidate();
        }

        @Override
        public void onTapCommand(int tapType, int x, int y, long eventTime) {
            switch (tapType) {
                case TAP_TYPE_TAP:
                    int tappedComplicationId = getTappedComplicationId(x, y);
                    if (tappedComplicationId != -1) {
                        onComplicationTap(tappedComplicationId);
                    }
                    break;
            }
        }

        private int getTappedComplicationId(int x, int y) {
            ComplicationData complicationData;
            long currentTimeMillis = System.currentTimeMillis();

            for (int i = 0; i < Constants.COMPLICATION_IDS.length; i++) {
                complicationData = mActiveComplicationDataSparseArray.get(Constants.COMPLICATION_IDS[i]);

                if ((complicationData != null)
                        && (complicationData.isActive(currentTimeMillis))
                        && (complicationData.getType() != ComplicationData.TYPE_NOT_CONFIGURED)
                        && (complicationData.getType() != ComplicationData.TYPE_EMPTY)) {

                    Rect complicationBoundingRect = new Rect(0, 0, 0, 0);

                    switch (Constants.COMPLICATION_IDS[i]) {
                        case Constants.LEFT_DIAL_COMPLICATION:
                            complicationBoundingRect.set(
                                    0,                                          // left
                                    mComplicationsY - Constants.COMPLICATION_TAP_BUFFER,  // top
                                    (mWidth / 2),                               // right
                                    ((int) Constants.COMPLICATION_TEXT_SIZE               // bottom
                                            + mComplicationsY
                                            + Constants.COMPLICATION_TAP_BUFFER));
                            break;

                        case Constants.RIGHT_DIAL_COMPLICATION:
                            complicationBoundingRect.set(
                                    (mWidth / 2),                               // left
                                    mComplicationsY - Constants.COMPLICATION_TAP_BUFFER,  // top
                                    mWidth,                                     // right
                                    ((int) Constants.COMPLICATION_TEXT_SIZE               // bottom
                                            + mComplicationsY
                                            + Constants.COMPLICATION_TAP_BUFFER));
                            break;
                    }

                    if (complicationBoundingRect.width() > 0) {
                        if (complicationBoundingRect.contains(x, y)) {
                            return Constants.COMPLICATION_IDS[i];
                        }
                    } else {
                        Log.e(TAG, "Not a recognized complication id.");
                    }
                }
            }
            return -1;
        }

        /*
         * Fires PendingIntent associated with complication (if it has one).
         */
        private void onComplicationTap(int complicationId) {
            ComplicationData complicationData = mActiveComplicationDataSparseArray.get(complicationId);

            if (complicationData != null) {
                if (complicationData.getTapAction() != null) {
                    try {
                        complicationData.getTapAction().send();
                    } catch (PendingIntent.CanceledException e) {
                        Log.e(TAG, "On complication tap action error " + e);
                    }
                } else if (complicationData.getType() == ComplicationData.TYPE_NO_PERMISSION) {

                    // Watch face does not have permission to receive complication data, so launch
                    // permission request.
                    ComponentName componentName = new ComponentName(
                            getApplicationContext(),
                            SlateWatchFaceService.class);

                    Intent permissionRequestIntent =
                            ComplicationHelperActivity.createPermissionRequestHelperIntent(
                                    getApplicationContext(), componentName);

                    startActivity(permissionRequestIntent);
                }

            }
        }

        @Override
        public void onAmbientModeChanged(boolean inAmbientMode) {
            /* the wearable switched between modes */
            super.onAmbientModeChanged(inAmbientMode);
            if (mLowBitAmbient) {
                mPaints.setAntiAlias(!inAmbientMode);
            }
            invalidate();
            restartTimer();
        }

        @Override
        public void onSurfaceChanged(SurfaceHolder holder, int format, int width, int height) {
            super.onSurfaceChanged(holder, format, width, height);

            mWidth = width;
            mHeight = height;

            /*
             * Since the height of the complications text does not change, we only have to
             * recalculate when the surface changes.
             */
            mComplicationsY = (int) ((mHeight / 2) + (mPaints.complicationPaint.getTextSize() / 3));

            initializeBackground(width, height);
            initializeTicks(width, height);
        }

        // Scale the background to fit.
        private void initializeBackground(int width, int height) {
            if (null == mBackgroundScaledBitmap || mBackgroundScaledBitmap.getWidth() != width || mBackgroundScaledBitmap.getHeight() != height) {
                mBackgroundScaledBitmap = Bitmap.createScaledBitmap(mBackgroundBitmap, width, height, true);
            }
        }

        private void initializeTicks(int width, int height) {
            float centerX = width / 2f;
            float centerY = height / 2f;

            float innerTickRadius = centerX - 18;
            float largeInnerTickRadius = centerX - 42;
            for (int tickIndex = 0; tickIndex < 12; tickIndex++) {
                float tickRot = (float) (tickIndex * Math.PI * 2 / 12);
                float innerX;
                float innerY;
                if (tickIndex == 0 || tickIndex == 3 || tickIndex == 6 || tickIndex == 9) {
                    innerX = (float) Math.sin(tickRot) * largeInnerTickRadius;
                    innerY = (float) -Math.cos(tickRot) * largeInnerTickRadius;
                } else {
                    innerX = (float) Math.sin(tickRot) * innerTickRadius;
                    innerY = (float) -Math.cos(tickRot) * innerTickRadius;
                }
                float outerX = (float) Math.sin(tickRot) * centerX;
                float outerY = (float) -Math.cos(tickRot) * centerX;
                mTicks[tickIndex] = new float[]{centerX + innerX, centerY + innerY, centerX + outerX, centerY + outerY};
            }
        }

        @Override
        public void onTimeTick() {
            super.onTimeTick();
            invalidate();
        }

        @Override
        public void onDraw(Canvas canvas, Rect bounds) {
            int width = bounds.width();
            int height = bounds.height();

            float centerX = width / 2f;
            float centerY = height / 2f;

            boolean isAmbient = isInAmbientMode();
            Calendar calendar = mSlateTime.getTimeNow();

            drawBackground(canvas, isAmbient);
            drawComplications(canvas, isAmbient, calendar.getTimeInMillis());
            drawTicks(canvas, isAmbient);
            drawHands(canvas, isAmbient, calendar, centerX, centerY);
        }

        private void drawBackground(Canvas canvas, boolean isAmbient) {
            if (isAmbient) {
                canvas.drawColor(Color.BLACK);
            } else {
                canvas.drawBitmap(mBackgroundScaledBitmap, 0, 0, null);
            }
        }

        private void drawComplications(Canvas canvas, boolean isAmbient, long currentTimeMillis) {
            ComplicationData complicationData;
            for (int i = 0; i < Constants.COMPLICATION_IDS.length; i++) {
                complicationData = mActiveComplicationDataSparseArray.get(Constants.COMPLICATION_IDS[i]);
                if (complicationData != null && complicationData.isActive(currentTimeMillis)) {

                    // Both Short Text and No Permission Types can be rendered with the same code.
                    // No Permission will display "--" with an Intent to launch a permission prompt.
                    // If you want to support more types, just add a "else if" below with your
                    // rendering code inside.
                    if (complicationData.getType() == ComplicationData.TYPE_SHORT_TEXT
                            || complicationData.getType() == ComplicationData.TYPE_NO_PERMISSION) {

                        ComplicationText mainText = complicationData.getShortText();
                        ComplicationText subText = complicationData.getShortTitle();

                        CharSequence complicationMessage =
                                mainText.getText(getApplicationContext(), currentTimeMillis);

                        /* In most cases you would want the subText (Title) under the
                         * mainText (Text), but to keep it simple for the code lab, we are
                         * concatenating them all on one line.
                         */
                        if (subText != null) {
                            complicationMessage = TextUtils.concat(complicationMessage, " ",
                                    subText.getText(getApplicationContext(), currentTimeMillis));
                        }

                        //Log.d(TAG, "Com id: " + COMPLICATION_IDS[i] + "\t" + complicationMessage);
                        double textWidth = mPaints.complicationPaint.measureText(
                                complicationMessage,
                                0,
                                complicationMessage.length());

                        int complicationsX;

                        if (Constants.COMPLICATION_IDS[i] == Constants.LEFT_DIAL_COMPLICATION) {
                            complicationsX = (int) ((mWidth / 2) - textWidth) / 2;
                        } else {
                            // RIGHT_DIAL_COMPLICATION calculations
                            int offset = (int) ((mWidth / 2) - textWidth) / 2;
                            complicationsX = (mWidth / 2) + offset;
                        }

                        float xOffset = (float) (textWidth / 2);
                        float radius = mHeight / 8;

                        if (!isAmbient) {
                            Paint temp = new Paint();
                            temp.setAntiAlias(true);
                            temp.setARGB(60, 0, 0, 0);

                            canvas.drawCircle(complicationsX + xOffset, mHeight / 2, radius, temp);

                            Paint temp2 = new Paint();
                            temp2.setAntiAlias(true);
                            temp2.setARGB(40, 80, 80, 80);

                            canvas.drawCircle(complicationsX + xOffset, mHeight / 2, radius - 4f, temp2);
                        }
                        canvas.drawText(
                                complicationMessage,
                                0,
                                complicationMessage.length(),
                                complicationsX,
                                mComplicationsY,
                                mPaints.complicationPaint);
                    }
                }
            }
        }

        private void drawTicks(Canvas canvas, boolean isAmbient) {
            if (!isAmbient) {
                for (float[] tick : mTicks) {
                    canvas.drawLines(tick, mPaints.tickPaint);
                }
            }
        }

        private void drawHands(Canvas canvas, boolean isAmbient, Calendar calendar, float centerX, float centerY) {
            Config config = mConfigService.getConfig();

            float milliRotate = ((calendar.getTimeInMillis() % 60000) / 1000f) / 30f * (float) Math.PI;
            float secRotate = calendar.get(Calendar.SECOND) / 30f * (float) Math.PI;
            int minutes = calendar.get(Calendar.MINUTE);
            float minRot = minutes / 30f * (float) Math.PI;
            float hrRot = ((calendar.get(Calendar.HOUR) + (minutes / 60f)) / 6f) * (float) Math.PI;

            float secLength = centerX - 16;
            float minLength = centerX - 26;
            float hrLength = centerX - 70;

            float hrX = (float) Math.sin(hrRot) * hrLength;
            float hrY = (float) -Math.cos(hrRot) * hrLength;
            canvas.drawLine(centerX, centerY, centerX + hrX, centerY + hrY, mPaints.hourPaint);

            float minX = (float) Math.sin(minRot) * minLength;
            float minY = (float) -Math.cos(minRot) * minLength;
            canvas.drawCircle(centerX, centerY, 10f, mPaints.minutePaint);
            canvas.drawLine(centerX, centerY, centerX + minX, centerY + minY, mPaints.minutePaint);
            canvas.drawCircle(centerX, centerY, 10f, mPaints.centerPaint);

            if (!isAmbient) {
                mPaints.setAccentColor(config.getAccentColor());

                float rotate = config.isSmoothMovement() ? milliRotate : secRotate;
                float secStartX = (float) Math.sin(rotate) * -40;
                float secStartY = (float) -Math.cos(rotate) * -40;
                float secX = (float) Math.sin(rotate) * secLength;
                float secY = (float) -Math.cos(rotate) * secLength;

                canvas.drawLine(centerX + secStartX,
                        centerY + secStartY,
                        centerX + secX,
                        centerY + secY, mPaints.secondPaint);

                canvas.drawCircle(centerX, centerY, 6f, mPaints.secondPaint);
            }
        }

        /**
         * The watch face became visible or invisible
         *
         * @param visible If the watch face is visible or not
         */
        @Override
        public void onVisibilityChanged(boolean visible) {
            super.onVisibilityChanged(visible);
            if (visible) {
                mConfigService.connect();
                mSlateTime.registerReceiver();
                mSlateTime.reset();
            } else {
                mSlateTime.unregisterReceiver();
                mConfigService.disconnect();
            }
            restartTimer();
        }

        /**
         * Whether the timer should be running depends on whether we're in ambient mode
         * (as well as whether we're visible), so we may need to start or stop the timer.
         * <p>
         * Starts the {@link #mUpdateTimeHandler} timer if it should be running and isn't currently
         * or stops it if it shouldn't be running but currently is.
         */
        private void restartTimer() {
            mUpdateTimeHandler.removeMessages(Constants.MSG_UPDATE_TIME);
            if (shouldTimerBeRunning()) {
                mUpdateTimeHandler.sendEmptyMessage(Constants.MSG_UPDATE_TIME);
            }
        }

        @Override
        public void onDestroy() {
            mUpdateTimeHandler.removeMessages(Constants.MSG_UPDATE_TIME);
            super.onDestroy();
        }

        /**
         * Returns whether the {@link #mUpdateTimeHandler} timer should be running. The timer should
         * only run when we're visible and in interactive mode.
         */
        private boolean shouldTimerBeRunning() {
            return isVisible() && !isInAmbientMode();
        }
    }

    private static class EngineHandler extends Handler {

        private final Engine mEngine;

        EngineHandler(Engine engine) {
            mEngine = engine;
        }

        @Override
        public void handleMessage(Message message) {
            switch (message.what) {
                case Constants.MSG_UPDATE_TIME:
                    mEngine.invalidate();
                    if (mEngine.shouldTimerBeRunning()) {
                        long updateRate = mEngine.mConfigService.getConfig().getUpdateRate();
                        long delayMs = updateRate - (System.currentTimeMillis() % updateRate);
                        mEngine.mUpdateTimeHandler.sendEmptyMessageDelayed(Constants.MSG_UPDATE_TIME, delayMs);
                    }
                    break;
            }
        }
    }
}

File: wear/src/main/java/ca/joshstagg/slate/SlateWatchFaceService.kt
package ca.joshstagg.slate

import android.content.ComponentName
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Rect
import android.os.Bundle
import android.os.Handler
import android.os.Message
import android.support.wearable.complications.ComplicationData
import android.support.wearable.watchface.CanvasWatchFaceService
import android.support.wearable.watchface.WatchFaceService
import android.support.wearable.watchface.WatchFaceStyle
import android.view.Gravity
import android.view.SurfaceHolder
import ca.joshstagg.slate.complication.ComplicationEngine
import java.util.*

/**
 * Slate ca.joshstagg.slate
 * Copyright 2017 Josh Stagg
 */
class SlateWatchFaceService : CanvasWatchFaceService() {

    override fun onCreateEngine(): Engine {
        return Engine(applicationContext)
    }

    inner class Engine internal constructor(private val context: Context) :
        CanvasWatchFaceService.Engine() {
        private val updateTimeHandler: Handler = EngineHandler(this)
        private val paints: SlatePaints = SlatePaints(context)
        private val slateTime: SlateTime = SlateTime(context)
        private val random = Random()

        /**
         * Whether the display supports fewer bits for each color in ambient mode. When true, we
         * disable anti-aliasing in ambient mode.
         */
        private var lowBitAmbient: Boolean = false
        private var burnInProtection: Boolean = false
        private var ambientMode: Ambient = Ambient.NORMAL

        private lateinit var watchEngine: WatchEngine
        private lateinit var complicationEngine: ComplicationEngine
        private lateinit var notificationEngine: NotificationEngine

        /**
         * Initialize the watch face

         * @param holder SurfaceHolder that the watch face will use
         */
        override fun onCreate(holder: SurfaceHolder?) {
            super.onCreate(holder)
            setWatchFaceStyle(
                WatchFaceStyle.Builder
                    .forComponentName(ComponentName(context, SlateWatchFaceService::class.java))
                    .setStatusBarGravity(Gravity.CENTER_HORIZONTAL or Gravity.TOP)
                    .setAcceptsTapEvents(true)
                    .setAccentColor(paints.accentHandColor)
                    .setHideNotificationIndicator(true)
                    .setShowUnreadCountIndicator(false)
                    .build()
            )
            watchEngine = WatchEngine(context, paints)
            complicationEngine = ComplicationEngine(applicationContext, paints)
            notificationEngine = NotificationEngine(paints)
            setActiveComplications(*COMPLICATION_IDS)
        }

        override fun onPropertiesChanged(properties: Bundle) {
            super.onPropertiesChanged(properties)
            properties.apply {
                lowBitAmbient = getBoolean(WatchFaceService.PROPERTY_LOW_BIT_AMBIENT, false)
                burnInProtection = getBoolean(WatchFaceService.PROPERTY_BURN_IN_PROTECTION, false)
            }
        }

        /*
         * Called when there is updated data for a complication id.
         */
        override fun onComplicationDataUpdate(
            complicationId: Int,
            complicationData: ComplicationData?
        ) {
            // Adds/updates active complication data in the array.
            complicationEngine.dataUpdate(complicationId, complicationData)
            invalidate()
        }

        override fun onTapCommand(tapType: Int, x: Int, y: Int, eventTime: Long) {
            when (tapType) {
                WatchFaceService.TAP_TYPE_TAP -> complicationEngine.complicationTap(x, y)
            }
        }

        override fun onAmbientModeChanged(inAmbientMode: Boolean) {
            /* the wearable switched between modes */
            super.onAmbientModeChanged(inAmbientMode)
            val config = Slate.instance.configService.config

            ambientMode = when {
                inAmbientMode && lowBitAmbient && burnInProtection -> Ambient.AMBIENT_LOW_BIT_BURN_IN
                inAmbientMode && burnInProtection -> Ambient.AMBIENT_BURN_IN
                inAmbientMode && lowBitAmbient -> Ambient.AMBIENT_LOW_BIT
                inAmbientMode -> Ambient.AMBIENT
                else -> Ambient.NORMAL
            }
            if (lowBitAmbient) {
                paints.setAntiAlias(!inAmbientMode)
            }

            if (inAmbientMode) {
                paints.handColor = config.ambientColor
                paints.second.color = Color.BLACK
            } else {
                paints.handColor = paints.primaryHandColor
                paints.second.color = config.accentColor
            }

            invalidate()
            restartTimer()
        }

        override fun onSurfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {
            super.onSurfaceChanged(holder, format, width, height)
            watchEngine.initialize(width, height)
            complicationEngine.initialize(width, height)
        }

        override fun onUnreadCountChanged(count: Int) {
            if (notificationEngine.unreadCountChanged(count)) {
                invalidate()
            }
        }

        override fun onTimeTick() {
            super.onTimeTick()
            invalidate()
        }

        override fun onDraw(canvas: Canvas, bounds: Rect) {
            val width = bounds.width()
            val height = bounds.height()

            val centerX = width / 2f
            val centerY = height / 2f

            val calendar = slateTime.timeNow

            if (Ambient.AMBIENT_BURN_IN == ambientMode ||
                Ambient.AMBIENT_LOW_BIT_BURN_IN == ambientMode
            ) {
                val shift = paints.burnInShift
                val shift2 = paints.burnInShift / 2
                val offsetX = (random.nextFloat() * shift) - shift2
                val offsetY = (random.nextFloat() * shift) - shift2
                Logger.d("SlateWatchFaceService", "Translate: $offsetX, $offsetY")
                canvas.translate(offsetX, offsetY)
            }

            watchEngine.drawBackground(canvas, ambientMode)
            complicationEngine.drawComplications(canvas, ambientMode, calendar)
            watchEngine.drawTicks(canvas, ambientMode)
            watchEngine.drawHands(canvas, ambientMode, calendar, centerX, centerY)
            notificationEngine.drawUnreadIndicator(canvas)
        }

        /**
         * The watch face became visible or invisible

         * @param visible If the watch face is visible or not
         */
        override fun onVisibilityChanged(visible: Boolean) {
            super.onVisibilityChanged(visible)
            if (visible) {
                Slate.instance.configService.connect()
                slateTime.registerReceiver()
                slateTime.reset()
            } else {
                slateTime.unregisterReceiver()
                Slate.instance.configService.disconnect()
            }
            restartTimer()
        }

        /**
         * Whether the timer should be running depends on whether we're in ambient mode
         * (as well as whether we're visible), so we may need to start or stop the timer.
         *
         *
         * Starts the [.updateTimeHandler] timer if it should be running and isn't currently
         * or stops it if it shouldn't be running but currently is.
         */
        private fun restartTimer() {
            updateTimeHandler.removeMessages(MSG_UPDATE_TIME)
            if (shouldTimerBeRunning()) {
                updateTimeHandler.sendEmptyMessage(MSG_UPDATE_TIME)
            }
        }

        override fun onDestroy() {
            updateTimeHandler.removeMessages(MSG_UPDATE_TIME)
            super.onDestroy()
        }

        /**
         * Returns whether the [.updateTimeHandler] timer should be running. The timer should
         * only run when we're visible and in interactive mode.
         */
        private fun shouldTimerBeRunning(): Boolean {
            return isVisible && !isInAmbientMode
        }

        fun updateTime() {
            invalidate()
            if (shouldTimerBeRunning()) {
                val config = Slate.instance.configService.config
                val updateRate = config.updateRate
                val delayMs = updateRate - System.currentTimeMillis() % updateRate
                updateTimeHandler.sendEmptyMessageDelayed(MSG_UPDATE_TIME, delayMs)
            }
        }
    }

    private class EngineHandler internal constructor(private val engine: Engine) : Handler() {
        override fun handleMessage(message: Message) {
            if (MSG_UPDATE_TIME == message.what) {
                engine.updateTime()
            }
        }
    }
}
